chain/xrp/tx/tx.go

```diff
@@ -0,0 +1,229 @@
+	if err != nil {
+		return xc.TxHash("")
+	}
+
+	hash := sha512.Sum512(decodedBytes)
+	firstHalf := hash[:32]
+	hashHex := hex.EncodeToString(firstHalf)
+
+	return xc.TxHash(hashHex)
+}
+
+// Sighashes returns the tx payload to sign, aka sighash
+func (tx Tx) Sighashes() ([]xc.TxDataToSign, error) {
+	if tx.XRPTx == nil {
+		return nil, errors.New("missing XRP transaction")
+	}
+
+	if tx.EncodeForSigning == nil {
+		return nil, errors.New("missing serialised XRP transaction")
+	}
+
+	// For k256 signing, XRP uses sha512[:32]
+	// https://github.com/XRPLF/xrpl-py/blob/17aad31f77452d30917b9e4544c9c87c274c0e3d/xrpl/core/keypairs/secp256k1.py#L95
+	digestSha512 := sha512.Sum512(tx.EncodeForSigning)
+	firstHalf := digestSha512[:32]
+
+	return []xc.TxDataToSign{firstHalf[:]}, nil
+
+}
+
+// AddSignatures adds a signature to Tx
+func (tx *Tx) AddSignatures(signatures ...xc.TxSignature) error {
+	if tx.TransactionSignature != nil {
+		return errors.New("transaction already signed")
+	}
+
+	for _, rsvBytes := range signatures {
+		r, s, err := btctx.DecodeEcdsaSignature(rsvBytes)
+		if err != nil {
+			return err
+		}
+
+		signature := ecdsa.NewSignature(&r, &s)
+		signatureBytes := signature.Serialize()
+		signatureHex := hex.EncodeToString(signatureBytes)
+		tx.XRPTx.TxnSignature = signatureHex
+		tx.TransactionSignature = append(tx.TransactionSignature, rsvBytes)
+	}
+
+	return nil
+}
+
+// GetSignatures returns back signatures, which may be used for signed-transaction broadcasting
+func (tx *Tx) GetSignatures() []xc.TxSignature {
+	return tx.TransactionSignature
+}
+
+// Serialize returns the serialized tx
+func (tx Tx) Serialize() ([]byte, error) {
+
+	if tx.XRPTx == nil {
+		return []byte{}, errors.New("missing XRP transaction")
+	}
+
+	xrpTx := tx.XRPTx
+
+	if xrpTx.LastLedgerSequence == 0 {
+		return []byte{}, errors.New("missing last ledger sequence")
+	}
+
+	if xrpTx.TxnSignature == "" {
+		return []byte{}, errors.New("missing transaction signature")
+	}
+
+	resultMapXRP := RenderToMap(*xrpTx)
+	resultMapWithAmount := make(map[string]interface{})
+
+	if xrpTx.Amount.IsString {
+		resultMapWithAmount = WithTokenAmount(resultMapXRP, xrpTx.Amount.StringValue)
+	} else {
+		resultMapWithAmount = WithTokenAmount(resultMapXRP, xrpTx.Amount.AmountValue)
+	}
+
+	encodeTx, err := binarycodec.Encode(resultMapWithAmount)
+	if err != nil {
+		return []byte{}, fmt.Errorf("failed to serialise serialised XRP transaction: %v", err)
+	}
+
+	encodeTxWithPrefix := TRANSACTION_HASH_PREFIX + string(encodeTx)
+
+	decodedBytes, err := hex.DecodeString(encodeTxWithPrefix)
+
+	hash := sha512.Sum512(decodedBytes)
+
+	firstHalf := hash[:32]
+	hashHex := hex.EncodeToString(firstHalf)
+	fmt.Println(hashHex)
+
+	return []byte(encodeTx), nil
+}
+
+func RenderToMap(xrpTx XRPTransaction) map[string]interface{} {
+	result := make(map[string]interface{})
+	result["Account"] = string(xrpTx.Account)
+	result["Amount"] = xrpTx.Amount
```

Comment:
I notice you later call `WithTokenAmount` after `RenderToMap`?  I think it'd be better if `RenderToMap` did the rendering of Amount correctly.